File: conversions/chunk-array/elide-chunk-array.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-chunk-array

This demonstrates calling the TypeScript chunk implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One chunk library shared across TypeScript and Python
- Consistent array chunking across services
- No need for custom Python chunk logic
- Guaranteed consistent batch handling
"""

print("=== Python Calling TypeScript Chunk Array ===\n")

# Example 1: Batch processing
# data = list(range(100))
# batches = chunk.default(data, 10)  # [[0-9], [10-19], ...]

# Example 2: API Pagination
# def paginate(items: list, page_size: int = 20) -> list[list]:
#     """Paginate items for API responses"""
#     return chunk.default(items, page_size)

# Example 3: Parallel Processing
# def process_in_batches(items: list, batch_size: int = 50):
#     """Process items in parallel batches"""
#     batches = chunk.default(items, batch_size)
#     return [process_batch(batch) for batch in batches]

print("Real-world use case:")
print("- Python data pipeline chunks data for batch processing")
print("- Uses same TypeScript implementation as Node.js API")
print("- Guarantees consistent batch sizes")
print("- No custom Python chunk logic needed")
print()

print("Example: Batch Processing Pipeline")
print("┌─────────────────────────────────────────────┐")
print("│   Elide Chunk Array (TypeScript)          │")
print("│   elide-chunk-array.ts                     │")
print("└─────────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │Pipeline│")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same batch chunking everywhere!")
print()

print("Problem Solved:")
print("Before: Different chunk implementations = inconsistent batch sizes")
print("After: One Elide implementation = consistent chunking")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  chunk = require('./elide-chunk-array.ts')")
print("  batches = chunk.default(data, 10)  # That's it!")

File: conversions/chunk-array/elide-chunk-array.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-chunk-array

puts "=== Ruby Calling TypeScript Chunk Array ===\n"

# Example 1: Rails Batch Processing
# class BatchProcessor
#   def process_in_batches(items, batch_size = 50)
#     batches = chunk_module.default(items, batch_size)
#     batches.each { |batch| process_batch(batch) }
#   end
# end

# Example 2: Pagination
# class ProductsController < ApplicationController
#   def index
#     all_products = Product.all
#     pages = chunk_module.default(all_products, 20)
#     render json: pages[params[:page].to_i]
#   end
# end

puts "Real-world use case:"
puts "- Ruby Rails app chunks data for batch processing"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent batch sizes"
puts "- No Array.each_slice needed"
puts ""

puts "Example: Batch Processing System"
puts "┌─────────────────────────────────────────────┐"
puts "│   Elide Chunk Array (TypeScript)          │"
puts "│   elide-chunk-array.ts                     │"
puts "└─────────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same chunking logic everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different chunk implementations = batch size inconsistencies"
puts "After: One Elide implementation = unified chunking"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/chunk.rb"
puts "  CHUNK = Elide.require('./elide-chunk-array.ts')"
puts "  "
puts "  # app/services/batch_processor.rb"
puts "  def chunk_data(data, size)"
puts "    CHUNK.default(data, size)"
puts "  end"
